rcl/include/rcl/event.h

```diff
@@ -0,0 +1,149 @@
+// Copyright 2015 Open Source Robotics Foundation, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef RCL__EVENT_H_
+#define RCL__EVENT_H_
+
+#ifdef __cplusplus
+extern "C"
+{
+#endif
+
+#include "rcl/client.h"
+#include "rcl/macros.h"
+#include "rcl/publisher.h"
+#include "rcl/service.h"
+#include "rcl/subscription.h"
+#include "rcl/visibility_control.h"
+
+
+typedef enum rcl_publisher_event_type_t
+{
+  RCL_PUBLISHER_DEADLINE,
+  RCL_PUBLISHER_LIVELINESS
+} rcl_publisher_event_type_t;
+
+
+typedef enum rcl_subscription_event_type_t
+{
+  RCL_SUBSCRIPTION_DEADLINE,
+  RCL_SUBSCRIPTION_LIVELINESS
+} rcl_subscription_event_type_t;
```

Comment:
Why are these defined here, but not stored in the event type directly? It seems to be in the impl or the rmw version of the event? If in the impl, then why not in the non-impl part of the struct, and if in rmw, why redefine the enum's here?

Reply:
`rclcpp` needs to pass in these types during `rcl_*_event_init()`. My understanding is that if these enums are strictly in the impl, then they will be unavailable to `rclcpp`.

Reply:
Ok I understand they are passed as arguments to the init functions, but why is the storage for them in the impl though? Why not as members of `rcl_event_t`. If it must stay in the `impl`, then why not have accessors functions for the event types?

If I have an event, it would be nice to know what kind by introspection. This migh imply that the event type enums need to be merged, into just "event types" and not subscription/publisher event types.

Reply:
These types (technically a translation of it to a `rmw_event_type_t` enum) are stored in `rmw_event_t`.

The `rclcpp` layer is aware of the event type because it is the layer setting it. The only other layer that actually needs to know the event type of a `rcl_event_t`/`rmw_event_t` is `rmw_fastrtps`/`rmw_connext`/`rmw_opensplice`. The event type just passes through the `rcl` and `rmw` layers.

You are right that the event type enums should be merged at some point. A merge does occur when translating `rcl_subscription_event_type_t` or `rcl_publisher_event_type_t` to `rmw_event_type_t`. We just thought that having separate `rcl_subscription_event_type_t`/`rcl_publisher_event_type_t` types is better for type-safety when calling the `rcl_*_event_init()` functions.

Reply:
Ok, that makes more sense to me now. I'll defer to you guys on what to change if anything. You don't have to merge them at my request, I didn't consider everything until you put it into context there.